cou/steps/plan.py

```diff
@@ -435,39 +447,84 @@ def _generate_control_plane_plan(
+    non_hypervisors_apps = []
+    for app in apps_data_plane:
+        if app.charm == "ceph-osd" or app.is_subordinate:
+            non_hypervisors_apps.append(app)
+            continue
+
+        hypervisor_apps.append(app)
+    return hypervisor_apps, non_hypervisors_apps
+
+
+async def _generate_hypervisors_plan(
     args: CLIargs,
-) -> list[UpgradePlan]:  # pragma: no cover
-    """Generate upgrade plan for data plane.
+    analysis_result: Analysis,
+    target: OpenStackRelease,
+    hypervisor_apps: list[OpenStackApplication],
+) -> UpgradePlan:
+    """Generate upgrade plan for hypervisors.
 
-    :param target: Target OpenStack release.
-    :type target: OpenStackRelease
-    :param analysis_result: Analysis result.
-    :type analysis_result: Analysis
     :param args: CLI arguments
     :type args: CLIargs
-    :return: Hypervisors plans and other data plane applications, e.g. ceph-osd
-    :rtype: list[UpgradePlan]
+    :param analysis_result: Analysis result
+    :type analysis_result: Analysis
+    :param target: Target OpenStack release.
+    :type target: OpenStackRelease
+    :param hypervisor_apps: Hypervisor apps
+    :type hypervisor_apps: list[OpenStackApplication]
+    :return: Hypervisors upgrade plan.
+    :rtype: UpgradePlan
     """
     hypervisors_machines = await filter_hypervisors_machines(args, analysis_result)
     logger.info("Hypervisors selected: %s", hypervisors_machines)
+    hypervisor_planner = HypervisorUpgradePlanner(hypervisor_apps, hypervisors_machines)
+    hypervisor_plan = hypervisor_planner.generate_upgrade_plan(target, args.force)
+    logger.debug("Generation of the hypervisors upgrade plan completed")
+    return hypervisor_plan
 
-    hypervisor_apps = []
-    for app in analysis_result.apps_data_plane:
-        if app.charm == "ceph-osd":
-            # ceph-osd will not be handled by hypervisor planner
-            continue
-
-        hypervisor_apps.append(app)
 
-    hypervisor_planner = HypervisorUpgradePlanner(hypervisor_apps, hypervisors_machines)
-    hypervisor_plans = hypervisor_planner.generate_upgrade_plan(target, args.force)
-    logger.debug("Generation of the hypervisors upgrade plan complete")
+def _generate_ceph_osd_and_subordinates_plan(
+    target: OpenStackRelease, non_hypervisors_apps: list[OpenStackApplication], force: bool
+) -> list[UpgradePlan]:
+    """Generate upgrade plan for ceph-osd and data-plane subordinates.
 
-    # generate the plan for the remaining data-plane apps
-    return [hypervisor_plans]
+    :param target:  Target OpenStack release.
+    :type target: OpenStackRelease
+    :param non_hypervisors_apps:  Non-hypervisor apps
+    :type non_hypervisors_apps: list[OpenStackApplication]
+    :param force: Whether the plan generation should be forced
+    :type force: bool
+    :return: list containing the upgrade plan for ceph-osd and subordinates
+    :rtype: list[UpgradePlan]
+    """
+    return [
+        create_upgrade_group(
```

Comment:
What is the benefit of using this helper function create_upgrade_group? Is that even right for ceph-osd?
We will have `Data Plane ceph-osd upgrade plan` step with only with sub steps `Upgrade plan for 'ceph-osd' to <target>`. 
Can we simply use
```python
for app in non_hypervisors_apps:
    app.generate_upgrade_plan
```

Reply:
The benefit of using `create_upgrade_group` is that it handles `HaltUpgradePlanGeneration` and also some logs.

Reply:
It would not be better if the app itself managed the HaltUpgradePlanGeneration? The generate plan function cat catch that, log that and return empty plan. Like that it would be much cleaner, what do you say?

Reply:
I agree, this PR is not for it. I opened the issue #312.